From a5b79a411b114a6eee9df91c04e07451e6044146 Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Thu, 13 Nov 2014 15:31:27 +0800
Subject: [PATCH 0965/1074] MLK-9832-1 power: imx6_usb_charger: add
 usb_charger_is_present

usb_charger_is_present is called when the usb charger is recognized
or the usb charger is disconnected.

Besides, it fixes a bug that the SDP does not call power_supply_changed.

Signed-off-by: Peter Chen <peter.chen@freescale.com>
[Original patch taken from git://git.freescale.com/imx/linux-2.6-imx.git]
Signed-off-by: Biyao Zhai <biyao.zhai@windriver.com>
---
 drivers/power/imx6_usb_charger.c |   24 +++++++++++++++++-------
 1 files changed, 17 insertions(+), 7 deletions(-)

diff --git a/drivers/power/imx6_usb_charger.c b/drivers/power/imx6_usb_charger.c
index 44bbc9f..680c8df 100644
--- a/drivers/power/imx6_usb_charger.c
+++ b/drivers/power/imx6_usb_charger.c
@@ -144,6 +144,16 @@ static int imx6_usb_charger_detect(struct usb_charger *charger)
 	return 0;
 }
 
+static void usb_charger_is_present(struct usb_charger *charger, bool present)
+{
+	if (present)
+		charger->present = 1;
+	else
+		charger->present = 0;
+
+	power_supply_changed(&charger->psy);
+}
+
 /*
  * imx6_usb_vbus_connect - inform about VBUS connection
  * @charger: the usb charger
@@ -162,12 +172,14 @@ int imx6_usb_vbus_connect(struct usb_charger *charger)
 
 	/* Start the 1st period charger detection. */
 	ret = imx6_usb_charger_detect(charger);
-	if (ret)
+	if (ret) {
 		dev_err(charger->dev,
 				"Error occurs during detection: %d\n",
 				ret);
-	else
-		charger->present = 1;
+	} else {
+		if (charger->psy.type == POWER_SUPPLY_TYPE_USB)
+			usb_charger_is_present(charger, true);
+	}
 
 	mutex_unlock(&charger->lock);
 
@@ -199,8 +211,7 @@ int imx6_usb_charger_detect_post(struct usb_charger *charger)
 		charger->max_current = 900;
 	}
 
-	power_supply_changed(&charger->psy);
-
+	usb_charger_is_present(charger, true);
 	mutex_unlock(&charger->lock);
 
 	return 0;
@@ -217,11 +228,10 @@ EXPORT_SYMBOL(imx6_usb_charger_detect_post);
 int imx6_usb_vbus_disconnect(struct usb_charger *charger)
 {
 	charger->online = 0;
-	charger->present = 0;
 	charger->max_current = 0;
 	charger->psy.type = POWER_SUPPLY_TYPE_MAINS;
 
-	power_supply_changed(&charger->psy);
+	usb_charger_is_present(charger, false);
 
 	return 0;
 }
-- 
1.7.5.4

